File: 2024/day-18/main.dart
import 'dart:io';
import 'dart:collection';

typedef Point = List<int>;

List<Point> parseInput(String fileName) {
  final input = File(fileName).readAsLinesSync();
  return input.map((line) {
    final coords = line.split(',').map(int.parse).toList();
    return [coords[0], coords[1]];
  }).toList();
}

bool isWithinBounds(int x, int y, int size) {
  return x >= 0 && x < size && y >= 0 && y < size;
}

bool bfs(Set<Point> grid, Point start, Point goal, int size) {
  final directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0]
  ];
  final queue = Queue<Point>();
  final visited = <Point>{};

  queue.add(start);
  visited.add(start);

  while (queue.isNotEmpty) {
    final current = queue.removeFirst();
    if (current[0] == goal[0] && current[1] == goal[1]) {
      return true;
    }

    for (final direction in directions) {
      final nx = current[0] + direction[0];
      final ny = current[1] + direction[1];

      if (isWithinBounds(nx, ny, size)) {
        final neighbor = [nx, ny];
        if (!grid.contains(neighbor) && visited.add(neighbor)) {
          queue.add(neighbor);
        }
      }
    }
  }
  return false;
}

int? findShortestPath(Set<Point> grid, Point start, Point goal, int size) {
  final directions = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0]
  ];
  final queue = Queue<MapEntry<Point, int>>();
  final visited = <Point>{};

  queue.add(MapEntry(start, 0));
  visited.add(start);

  while (queue.isNotEmpty) {
    final entry = queue.removeFirst();
    final current = entry.key;
    final steps = entry.value;

    if (current[0] == goal[0] && current[1] == goal[1]) {
      return steps;
    }

    for (final direction in directions) {
      final nx = current[0] + direction[0];
      final ny = current[1] + direction[1];

      if (isWithinBounds(nx, ny, size)) {
        final neighbor = [nx, ny];
        if (!grid.contains(neighbor) && visited.add(neighbor)) {
          queue.add(MapEntry(neighbor, steps + 1));
        }
      }
    }
  }
  return null;
}

void main() {
  const fileName = 'input.txt';
  const size = 71;
  final points = parseInput(fileName);
  final start = [0, 0];
  final goal = [70, 70];

  // Part 1: Shortest path after 1024 bytes
  final gridPart1 = <Point>{};
  for (int i = 0; i < points.length && i < 1024; i++) {
    gridPart1.add(points[i]);
  }
  final shortestPath = findShortestPath(gridPart1, start, goal, size);
  if (shortestPath != null) {
    print('Shortest Path after 1024 bytes: $shortestPath');
  } else {
    print('No valid path to the exit after 1024 bytes.');
  }

  // Part 2: First byte that blocks the path
  final gridPart2 = <Point>{};
  for (int i = 0; i < points.length; i++) {
    gridPart2.add(points[i]);
    if (!bfs(gridPart2, start, goal, size)) {
      print('First blocking byte: ${points[i][0]},${points[i][1]}');
      break;
    }
  }
}

File: 2024/day-18/main.ex
defmodule MemoryGrid do
  def parse_input(file_name) do
    file_name
    |> File.read!()
    |> String.split("\n", trim: true)
    |> Enum.map(fn line ->
      line
      |> String.split(",")
      |> Enum.map(&String.to_integer/1)
    end)
  end

  defp is_within_bounds(x, y, size) do
    x >= 0 and x < size and y >= 0 and y < size
  end

  def bfs(grid, start, goal, size) do
    directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]
    queue = :queue.new() |> :queue.in(start)
    visited = MapSet.new([start])

    bfs_loop(queue, visited, directions, grid, goal, size)
  end

  defp bfs_loop(queue, visited, _directions, grid, goal, _size) do
    case :queue.out(queue) do
      {:empty, _} -> false
      {{:value, current}, new_queue} ->
        if current == goal, do: true, else: bfs_step(new_queue, visited, grid, goal, current, size)
    end
  end

  defp bfs_step(queue, visited, grid, goal, current, size) do
    directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]

    Enum.reduce_while(directions, queue, fn direction, acc ->
      [dx, dy] = direction
      [x, y] = current
      nx = x + dx
      ny = y + dy

      if is_within_bounds(nx, ny, size) do
        neighbor = [nx, ny]
        if !MapSet.member?(visited, neighbor) and !Enum.member?(grid, neighbor) do
          visited = MapSet.put(visited, neighbor)
          {:cont, :queue.in(neighbor, acc)}
        else
          {:cont, acc}
        end
      else
        {:cont, acc}
      end
    end)
    |> case do
      {:cont, new_queue} -> bfs_loop(new_queue, visited, grid, goal, size)
      _ -> false
    end
  end

  def find_shortest_path(grid, start, goal, size) do
    directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]
    queue = :queue.new() |> :queue.in({start, 0})
    visited = MapSet.new([start])

    find_shortest_path_loop(queue, visited, directions, grid, goal, size)
  end

  defp find_shortest_path_loop(queue, visited, directions, grid, goal, size) do
    case :queue.out(queue) do
      {:empty, _} -> nil
      {{:value, {{x, y}, steps}}, new_queue} ->
        if [x, y] == goal do
          steps
        else
          Enum.reduce_while(directions, new_queue, fn direction, acc ->
            [dx, dy] = direction
            nx = x + dx
            ny = y + dy
            if is_within_bounds(nx, ny, size) do
              neighbor = [nx, ny]
              if !MapSet.member?(visited, neighbor) and !Enum.member?(grid, neighbor) do
                visited = MapSet.put(visited, neighbor)
                {:cont, :queue.in({neighbor, steps + 1}, acc)}
              else
                {:cont, acc}
              end
            else
              {:cont, acc}
            end
          end)
          |> case do
            {:cont, new_queue} -> find_shortest_path_loop(new_queue, visited, directions, grid, goal, size)
            _ -> nil
          end
        end
    end
  end

  def solve do
    input_file = "input.txt"
    points = parse_input(input_file)
    size = 71
    start = [0, 0]
    goal = [70, 70]

    # Part 1: Shortest path after 1024 bytes
    grid_part1 = Enum.take(points, 1024) |> Enum.into(MapSet.new())
    case find_shortest_path(grid_part1, start, goal, size) do
      nil -> IO.puts("No valid path to the exit after 1024 bytes.")
      shortest_path -> IO.puts("Shortest Path after 1024 bytes: #{shortest_path}")
    end

    # Part 2: First byte that blocks the path
    grid_part2 = MapSet.new()
    Enum.each(points, fn point ->
      grid_part2 = MapSet.put(grid_part2, point)
      unless bfs(grid_part2, start, goal, size) do
        IO.puts("First blocking byte: #{Enum.join(point, ",")}")
        throw :done
      end
    end)
  end
end

MemoryGrid.solve()
